azurerm/resource_arm_iothub_resource_provider.go

```diff
@@ -0,0 +1,271 @@
+	name := d.Get("name").(string)
+
+	res, err := iothubClient.CheckNameAvailability(ctx, devices.OperationInputs{
+		Name: &name,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if !*res.NameAvailable {
+		_, err := iothubClient.Get(ctx, rg, name)
+		if err != nil {
+			return errors.New(string(res.Reason))
+		}
+	}
+
+	location := d.Get("location").(string)
+
+	subscriptionID := armClient.subscriptionId
+	skuInfo := expandAzureRmIotHubSku(d)
+
+	desc := devices.IotHubDescription{
+		Resourcegroup:  &rg,
+		Name:           &name,
+		Location:       &location,
+		Subscriptionid: &subscriptionID,
+		Sku:            &skuInfo,
+	}
+
+	if tagsI, ok := d.GetOk("tags"); ok {
+		tags := tagsI.(map[string]interface{})
+		desc.Tags = *expandTags(tags)
+	}
+
+	future, err := iothubClient.CreateOrUpdate(ctx, rg, name, desc, "")
+	if err != nil {
+		return err
+	}
+
+	err = future.WaitForCompletion(ctx, iothubClient.Client)
+	if err != nil {
+		return fmt.Errorf("Error creating or updating IotHub %q (Resource Group %q): %+v", name, rg, err)
+	}
+
+	desc, err = iothubClient.Get(ctx, rg, name)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*desc.ID)
+	return resourceArmIotHubRead(d, meta)
+
+}
+
+func expandAzureRmIotHubSku(d *schema.ResourceData) devices.IotHubSkuInfo {
+	skuList := d.Get("sku").([]interface{})
+	skuMap := skuList[0].(map[string]interface{})
+	cap := int64(skuMap["capacity"].(int))
+
+	name := skuMap["name"].(string)
+	tier := skuMap["tier"].(string)
+
+	return devices.IotHubSkuInfo{
+		Name:     devices.IotHubSku(name),
+		Tier:     devices.IotHubSkuTier(tier),
+		Capacity: &cap,
+	}
+
+}
+
+func resourceArmIotHubRead(d *schema.ResourceData, meta interface{}) error {
+	armClient := meta.(*ArmClient)
+	iothubClient := armClient.iothubResourceClient
+	ctx := armClient.StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	iothubName := id.Path["IotHubs"]
+	desc, err := iothubClient.Get(ctx, id.ResourceGroup, iothubName)
+
+	if err != nil {
+		return err
+	}
+
+	keysResp, err := iothubClient.ListKeys(ctx, id.ResourceGroup, iothubName)
+	keyList := keysResp.Response()
+
+	var keys []map[string]interface{}
+	for _, key := range *keyList.Value {
+		keyMap := make(map[string]interface{})
+
+		if keyName := key.KeyName; keyName != nil {
+			keyMap["key_name"] = *keyName
+		}
+
+		if primaryKey := key.PrimaryKey; primaryKey != nil {
+			keyMap["primary_key"] = *primaryKey
+		}
+
+		if secondaryKey := key.SecondaryKey; secondaryKey != nil {
+			keyMap["secondary_key"] = *secondaryKey
+		}
+
+		keyMap["permissions"] = string(key.Rights)
+		keys = append(keys, keyMap)
+	}
+
+	if err := d.Set("shared_access_policy", keys); err != nil {
+		return fmt.Errorf("Error flattening `shared_access_policy` in IoTHub %q: %+v", iothubName, err)
+	}
+
+	if properties := desc.Properties; properties != nil {
+		d.Set("hostname", properties.HostName)
+	}
+
+	d.Set("type", desc.Type)
+	flattenAndSetTags(d, &desc.Tags)
```

Comment:
so the Import test (`TestAccAzureRMIotHub_importBasic `) currently fails due to the following fields not being set - can we set them:

- `location`
- `name`
- `resource_group_name`
- `sku`
  - `capacity`
  - `name`
  - `tier`

on that note - could we pull this out into a new `flatten` function? I believe this should be what's needed to flatten the SKU block:

```
func flattenArmIoTHubSku(input *devices.IotHubSkuInfo) []interface {
  sku := make(map[string]interface{}, 0)

  if capacity := input.Capacity; capacity != nil {
    sku["capacity"] = *capacity
  }

  if name := input.Name; name != nil {
    sku["name"] = *name
  }

  if tier := input.Tier; tier != nil {
    sku["tier"] = *tier
  }

  return []interface{}{result}
}
```